### DISCUSSION TITLE
mapDoubleTap breaks the normal behavior (single tap, holding)

### DISCUSSION THREAD

**Discussion by mkjawadi** [evan-liu/karabiner.ts#158] [Q&A]
Hi, `mapDoubleTap` messes up the usual key behavior (single tap/holding). For example, below config maps that the double tapping the right shift key should toggle the “show desktop” option which works. However single-tap and hold actions for right shift don’t work. So, if you press the right shift key and a regular key (right shift + a), you won’t get the output you want, like typing the capital “A”. BTW, the double tap issue is with all keys not just the right shift key.

Karabiner Element: `v15.3.0`
karabiner.ts: `v1.30.3`

```ts
writeToProfile("Default profile", [
  rule("Caps Lock → Hyper").manipulators([
    map("⇪").toHyper().toIfAlone("⎋"), // Or map('caps_lock').toHyper().toIfAlone('escape')
    map("⇪", "⇧").to("⇪"), // Or map('caps_lock', 'shift').to('caps_lock')

    // Toggle Show Desktop
    mapDoubleTap("right_shift").to("f11", "fn")
  ])
]);
```
I have also tried the followings but didn't work either.
```ts
mapDoubleTap("right_shift").to("f11", "fn").toIfAlone("right_shift")
```
```ts
mapDoubleTap("right_shift").to("f11", "fn").singleTap(toKey("right_shift"))
```

Am I missing something?
Thanks in advance. 🙏

**Comment by evan-liu** [comment]
Try 

```ts
mapDoubleTap("right_shift").to("f11", "fn").toIfHeldDown("right_shift")
```

**Comment by mkjawadi** [comment]
Thanks for the response. No, it doesn't work unless I hold the shift for at least one second which is super slow specially while typing.

**Reply by evan-liu** [reply]
Try to set the parameters: 

- `basic.to_delayed_action_delay_milliseconds` for double tap
- `basic.to_if_held_down_threshold_milliseconds` for held down

```ts
mapDoubleTap("right_shift")
  .to("f11", "fn")
  .toIfHeldDown("right_shift")
  .parameters({
    "basic.to_delayed_action_delay_milliseconds": 500,
    "basic.to_if_held_down_threshold_milliseconds": 500
  })
```

**Reply by mkjawadi** [reply]
Sorry for the late response. I'd missed this one and unfortunately, it doesn’t work as expected, even after adjusting the milliseconds. Maybe I’m using the wrong option.

**Reply by evan-liu** [reply]
Sorry to hear that. I do not use double tap or layer (I use `duoLayer` instead) myself for this reason.

**Reply by situmorang-com** [reply]
this code:  
```
  rule('OSX: DoubleTaps').manipulators([
    mapDoubleTap('l⇧', '', 'any').to('⇪')
  ]),
```
  is creating this:
 ```
"to_if_invoked": [
                    { "key_code": "left_shift" },
```
   
   while I want this key_code line to be here like in this code:
```
{
    "description": "OSX: DoubleTaps",
    "manipulators": [
        {
            "conditions": [
                {
                    "name": "double-tap-left_shift-any",
                    "type": "variable_if",
                    "value": 1
                }
            ],
            "description": "__support__manipulator",
            "from": {
                "key_code": "left_shift",
                "modifiers": { "optional": ["any"] }
            },
            "to": [{ "key_code": "caps_lock" }],
            "type": "basic"
        },
        {
            "conditions": [
                {
                    "name": "double-tap-left_shift-any",
                    "type": "variable_unless",
                    "value": 1
                }
            ],
            "from": {
                "key_code": "left_shift",
                "modifiers": { "optional": ["any"] }
            },
            "parameters": { "basic.to_delayed_action_delay_milliseconds": 200 },
            "to": [
                {
                    "set_variable": {
                        "name": "double-tap-left_shift-any",
                        "value": 1
                    }
                },
                { "key_code": "left_shift" }
            ],
            "to_delayed_action": {
                "to_if_canceled": [
                    {
                        "set_variable": {
                            "name": "double-tap-left_shift-any",
                            "value": 0
                        }
                    }
                ],
                "to_if_invoked": [
                    {
                        "set_variable": {
                            "name": "double-tap-left_shift-any",
                            "value": 0
                        }
                    }
                ]
            },
            "type": "basic"
        }
    ]
}
```
Please help me navigate to create this because this works without the delay on shift key 😊🙏

**Reply by evan-liu** [reply]
Hi [@situmorang-com](https://github.com/situmorang-com) , can you please open a new discussion and explain more about what you need? It is not very clear on which part of the JSON you were talking about.

**Reply by thienanle** [reply]
@evan-liu I think the key part of @situmorang-com 's JSON is this:

```
            "to": [
                {
                    "set_variable": {
                        "name": "double-tap-left_shift-any",
                        "value": 1
                    }
                },
                { "key_code": "left_shift" }
            ],
```

This allows left_shift to work when held down. This is not desirable for keys that produce a character, but for modifiers it is better than toIfHeldDown as there is zero delay.

I suggest either it be the default behavior for modifiers, OR singleTap keyCode is inserted at [line 155 of double-tap.ts](https://github.com/evan-liu/karabiner.ts/blob/main/src/config/double-tap.ts#L155) if it is a modifier and the same as the double-tap keyCode.

If you'd like, I could try to create a PR, but wanted to discuss the best approach first.

**Reply by evan-liu** [reply]
Thanks, @thienanle. A PR would be great. Could you please open an issue to describe the problem and proposed solution, and we can discuss there?

**Reply by evan-liu** [reply]
I finally get what everyone is talking about. Sorry, I kept missing the issue is that modifiers need different behavior. @thienanle I am on it now. Thanks for offering the help.

**Reply by evan-liu** [reply]
Please upgrade to `1.34.1` and let me know if it works for you now.

**Comment by bboydflo** [comment]
Hey everyone!

First of all, thank you @evan-liu  for this awesome project! I am just getting started, but I have tried multiple things so far and my conclusion is that §§mapDoubleTap§§ is not working as intended.

I have tried a few things, among others, something like below: 

```
writeToProfile('default', [

  // tried to get this to work
  rule('Aliases', ifApp('Ghostty')).manipulators([
    mapDoubleTap('a')
      .to$(`aerospace list-windows --all | fzf --bind 'enter:execute(fish -c "aerospace focus --window-id {1}")+abort'`),
  ]),

  // also this one as a test
  rule('Double Tap of a').manipulators([
    mapDoubleTap('s').to$('ls -la'),
  ]),

])
```

And here are my `karabiner.json` parameters:

```
"parameters": {
      "basic.to_if_alone_timeout_milliseconds": 1000,
      "basic.to_if_held_down_threshold_milliseconds": 500,
      "basic.to_delayed_action_delay_milliseconds": 500,
      "basic.simultaneous_threshold_milliseconds": 50,
      "mouse_motion_to_scroll.speed": 100
}
```


Neither work for me. Am I missing something?

**Reply by evan-liu** [reply]
Hi @bboydflo, 

Can you try to confirm whether it is `mapDoubleTap()` or the command that is not working? 

1. Choose/create a simple mapping that is easy to verify. e.g. `map('f').toApp('Finder')`, and make sure it works. 
    1. Change `map()` to `mapDoubleTap()` and check if it works. e.g. `mapDoubleTap('f').toApp('Finder')`
    1. Change the action to a command you are using and check if it works. e.g. `map('f').to$('{yourCmdForDoubleTap}')`

**Reply by bboydflo** [reply]
Hi. I just tried `map('s').toApp('Finder')` and it does work. I have played with different variants and here are some results

```
rule('Double Tap1').manipulators([
  // map('f').toApp('Finder'), // does work
  mapDoubleTap('a').toApp('Finder'), // does work
]),

rule('Double Tap2', ifApp('com.mitchellh.ghostty')).manipulators([
    // map('x').toApp('Finder'), // works
    map('x').to$('ls -la'), // does not work
    // mapDoubleTap('x').to$('ls -la'), // does not work
]),

rule('Double Tap3', ifApp('Ghostty')).manipulators([
    map('x').toApp('Finder'), // does not work
]),
```
I learned that `ifApp()` works if I use the bundle identifier instead of the app name. But `to$()` will still not work. Tried with different terminals and shells.

I am using karabiner-elements 15.3.0

**Reply by bboydflo** [reply]
I have a suspicion about what's going on. Issue might be related to my default shell which is fish :( I've seen some logs in karabiner-elements and it tries to run a command with /bin/sh but I am pretty sure that the tool can't be run from /bin/sh.

Although, i've tried other commands as well, such as: `ls -la`

**Reply by evan-liu** [reply]
I think what you need is [`toPaste()`](https://github.com/evan-liu/karabiner.ts/blob/47459fc2fd4cb509de163e6cac6690ed3e7d7294/src/config/to.ts#L110) which paste the command into your terminal, instead of running the command.

**Reply by bboydflo** [reply]
Thanks! It works. Sorry for the trouble! I probably misunderstand the documentation

**Comment by evan-liu** [comment]
`mapDoubleTap('left_shift')` now produce: 

```json5
{
  "description": "mapDoubleTap()",
  "manipulators": [
    {
      "type": "basic",
      "from": { "key_code": "left_shift" },
      "conditions": [
        { "type": "variable_if", "name": "double-tap-left_shift", "value": 1 }
      ],
      "description": "__support__manipulator"
    },
    {
      "type": "basic",
      "from": { "key_code": "left_shift" },
      "to": [
        { "set_variable": { "name": "double-tap-left_shift", "value": 1 } },
        { "key_code": "left_shift", "lazy": true } // 1 Quick roll over without holding down
      ],
      "conditions": [
        {
          "type": "variable_unless",
          "name": "double-tap-left_shift",
          "value": 1
        }
      ],
      "to_delayed_action": {
        "to_if_invoked": [
          { "key_code": "left_shift" }, // 2 Tap
          { "set_variable": { "name": "double-tap-left_shift", "value": 0 } }
        ],
        "to_if_canceled": [
          { "set_variable": { "name": "double-tap-left_shift", "value": 0 } }
        ]
      },
      "to_if_held_down": [{ "key_code": "left_shift" }], // 3 Held
      "parameters": {
        "basic.to_if_held_down_threshold_milliseconds": 200,
        "basic.to_delayed_action_delay_milliseconds": 200
      }
    }
  ]
}

```
